### ISSUE TITLE
Streamline process of assigning people to training events

### ISSUE THREAD

**Issue by gvwilson** [carpentries/amy#1001]
Current workflow:
1. Go to `/workshops/training_requests/`.
2. Search for a particular person.
3. Click on the pencil icon to edit that TrainingRequest record.
4. Match the TrainingRequest to a person _and_ mark the application as "Accepted".
5. Return to the list of training requests.
6. Select that person's record (tick box in the left column).
7. Select an upcoming training event from the pull down and submit in order to match that person to the training event.

There are several ways to streamline this:
1. When editing a particular TrainingRequest: as soon as the "Accepted" box is ticked, a pull-down should appear to allow the user to assign that person to an upcoming training event.
2. On the page that shows all TrainingRequests, the column title "Matched Trainee" should be a pulldown allowing me to match the person right there and then to an existing Person record, while the "State" column should allow me to change their state without leaving this page.

**Comment by chrismedrela** [comment]
> On the page that shows all TrainingRequests, the column title "Matched Trainee" should be a pulldown allowing me to match the person right there and then to an existing Person record, while the "State" column should allow me to change their state without leaving this page.

@gvwilson, you can accept a request and match it with person record (or create a new one) in detailed view. That is, now, you can match a request with an existing trainee account in two clicks: first click "i" icon and then click "accept & match to selected trainee account". You're automaticaly redirected back to list view. Is it quick enough?

> When editing a particular TrainingRequest: as soon as the "Accepted" box is ticked, a pull-down should appear to allow the user to assign that person to an upcoming training event.

After clicking "accept & match to selected trainee account", you need select a request (that is, check the checkbox), scroll down, select training and click "match selected trainees to chosen training". That is four actions.

To sum up, each request can be processed with 2 + 4 = 6 actions now. I don't see how this can be significantly improved except that we could suggest incoming training, so to make selecting a training faster.

**Comment by chrismedrela** [comment]
After the call with @gvwilson, I agree with the proposed improvements.

**Comment by chrismedrela** [comment]
> On the page that shows all TrainingRequests, the column title "Matched Trainee" should be a pulldown allowing me to match the person right there and then to an existing Person record, while the "State" column should allow me to change their state without leaving this page.

What if we provide only a button like `match with Bob Smith <bob.smith@example.com> account` (instead of a pulldown) only if there is exactly one matching account and no shortcut otherwise? Is it enough?

> When editing a particular TrainingRequest: as soon as the "Accepted" box is ticked, a pull-down should appear to allow the user to assign that person to an upcoming training event.

Could we leave "state" column uneditable and automatically mark a request accepted when you match it with a training?

You propose to add "submit" button at the bottom, so you can change multiple trainees. What if the view was reloaded after clicking at `match with Bob Smith <bob.smith@example.com> account` button and one more after selecting a training from a drop-down (or clicking explicit submit button at the right side of the drop-down)? That makes implementation much simpler (no need for formsets, validation is much simpler, we don't need JS), but this results in a lot of page refreshes (two refreshes for each request), which may be disruptive.

**Comment by gvwilson** [comment]
On 2016-09-18 1:42 PM, Christopher Medrela wrote:

>     On the page that shows all TrainingRequests, the column title
>     "Matched Trainee" should be a pulldown allowing me to match the
>     person right there and then to an existing Person record, while
>     the "State" column should allow me to change their state without
>     leaving this page.
>
> What if we provide only a button like |match with Bob Smith 
> <bob.smith@example.com> account| (instead of a pulldown) only if there 
> is exactly one matching account and no shortcut otherwise? Is it enough?
>

A significant number of the matches this time were by name, but with 
different email addresses, so I'd strongly prefer the pulldown.

>     When editing a particular TrainingRequest: as soon as the
>     "Accepted" box is ticked, a pull-down should appear to allow the
>     user to assign that person to an upcoming training event.
>
> Could we leave "state" column uneditable and automatically mark a 
> request accepted when you match it with a training?
>
Yes.

> You propose to add "submit" button at the bottom, so you can change 
> multiple trainees. What if the view was reloaded after clicking at 
> |match with Bob Smith <bob.smith@example.com> account| button and one 
> more after selecting a training from a drop-down (or clicking explicit 
> submit button at the right side of the drop-down)? That makes 
> implementation much simpler (no need for formsets, validation is much 
> simpler, we don't need JS), but this results in a lot of page 
> refreshes (two refreshes for each request), which may be disruptive.
>

I'd prefer to be able to do many matches at once.  I realize formsets 
are a pain, but all the back-end logic is there (I hope?)

Thanks,
Greg

**Comment by chrismedrela** [comment]
> A significant number of the matches this time were by name, but with
different email addresses, so I'd strongly prefer the pulldown.

AMY searches for an existing account by matching email **or name**.

**Comment by gvwilson** [comment]
So what happens if the name matches but the email address doesn't?  Or 
if there are several people with the same name (either distinct 
individuals, or because we have two records for the same person and have 
only just realized it)?  I strongly prefer a pulldown.

**Comment by chrismedrela** [comment]
> So what happens if the name matches but the email address doesn't?

If there is only one Person record with exactly the same name, AMY suggests this Person (I'm talking about the form in TrainingRequest detailed form where you can match the request with a Person). If there is more than one matching record, AMY doesn't guess and don't propose anything. Should I change the logic there?

**Comment by gvwilson** [comment]
Nope, please leave that logic as-is --- I would just like a pulldowns in 
the "Matched Trainee" and "Matched Training" columns of 
https://amy.software-carpentry.org/workshops/training_requests/ to make 
it easier to match and assign lots of people.

**Comment by maneesha** [comment]
Closing this because we have made significant changes in how we accept Training Requests over the past year or two.
